samples: peripheral: Add sample demonstrating how to offload application for to VPR (flpr, ppr) (#27644)
nrfconnect/sdk-nrf: samples/peripheral/vpr_offloading/README.rst

The sample periodically reports the CPU load of the application core.

Offloaded modules
=================

The sample implements two modules that can be offloaded:

* LED control - Blinks the LED using a specific pattern.
* Temperature monitor - Periodically reads the temperature from a sensor and reports when it exceeds the defined range.
  It also periodically reports the average temperature.
  The module can use the LPS22HH sensor.
  If the sensor is not detected, SPI transfers are still performed but with random values.
  The module uses the NRFX_SPIM driver in blocking mode since the transfers are very short.

Example performance results
***************************

The following table shows the results of offloading in a scenario where both modules run for 2.5 seconds followed by a 0.5-second sleep.
The temperature monitor performs an SPI transfer every 1 millisecond.
The LED control module is blinking the **LED0** using various patterns.
Results include CPU load on the application core and average current.

.. list-table::
   :header-rows: 1
   :widths: auto

   * - Target
     - No offloading
     - Offloading

>    * - nrf54h20

Reviewer: ```suggestion
   * - nRF54H20
```

>      - 3.2 %, 126 uA

Reviewer: ```suggestion
     - 3.2 %, 126 µA
```

>      - 0.1 %, 71 uA

Reviewer: ```suggestion
     - 0.1 %, 71 µA
```

>    * - nrf54l15

Reviewer: ```suggestion
   * - nRF54L15
```

     - 3.0 %, 146 uA

>      - 0.1 %, 125 uA

Reviewer: ```suggestion
     - 3.0 %, 146 µA
     - 0.1 %, 125 µA
```

>    * - nrf54lm20a

Reviewer: ```suggestion
   * - nRF54LM20A
```

     - 3.3 %, 217 uA

>      - 0.0 %, 217 uA

Reviewer: ```suggestion
     - 3.3 %, 217 µA
     - 0.0 %, 217 µA
```

>    * - nrf54lv10a

Reviewer: ```suggestion
   * - nRF54LV10A
```

     - 3.3 %, 228 uA

>      - 0.0 %, 194 uA

Reviewer: ```suggestion
     - 3.3 %, 228 µA
     - 0.0 %, 194 µA
```

.. _vpr_offloading_sample:

VPR offloading
##############

.. contents::
   :local:
   :depth: 2

The VPR offloading sample demonstrates how VPR (PPR or FLPR) core can be used to offload the application core.

Overview
********

VPR cores can be used to offload the application core with background tasks.
It can reduce CPU load of the main core as well as power consumption.
VPR cores execute code from RAM so NVM memory does not need to be woken up.
Additionally, PPR on nRF54H20 runs in low power domain that is clocked with 16 MHz clock so power consumption savings can be significant compared to the application core running 320 MHz.

> On the other hand, VPR code space is limited so a lightweight framework for offloading is needed.

Reviewer 2: ```suggestion
VPR cores execute code from RAM, which means non-volatile memory (NVM) does not need to be woken up.
On the nRF54H20 SoC, PPR runs in the low-power domain clocked at 16 MHz, so the power savings can be significant compared to the application core running at 320 MHz.
However, the VPR code space is limited, so it requires a lightweight offloading framework.
```

  Module can use ``LPS22HH`` sensor.
  If sensor is not detected then SPI transfers are still performed but random values are used.
  Module is using ``nrfx_spim`` driver in the blocking mode since transfers are very short.

Depending on the configuration, modules can be executed on the application or the VPR core.
If the VPR core is the executor then the application core is using the IPC service.
Sample periodically reports the CPU load of the application core.

Table shows the results of the offloading in scenario where both modules run for 2,5 seconds followed by 0,5 second sleep.
Temperature monitor is performing SPI transfer every 1 ms. Results include CPU load on the application core and average current.

==================  =============  =============
target              no offloading  offloading
==================  =============  =============
nrf54h20 cpuppr     3.2 %, 126 uA  0.1 %, 71 uA
nrf54lm20a cpuflpr  3.3 %, 217 uA  0.0 %, 217 uA
nrf54lv10a cpuflpr  3.3 %, 228 uA  0.0 %, 194 uA
==================  =============  =============

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

> 

Author: it's just blinking various patterns and LED0 is controlled by VPR core while APP is sleeping. Can you take a look if current description is sufficient.
